ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs

```diff
@@ -338,30 +354,299 @@ toBearerInfo abi =
+          $ \clientAsync -> do
+            _ <- listenAndConnect (TestAddress 1) (TestAddress 0)
+                                 snocket getUnivState
+            wait clientAsync
 
 --
 -- Utils
 --
 
+runServer
+  :: ( MonadThread m
+     , MonadThrow m
+     )
+  => TestAddress addr -- ^ Local Address
+  -> Snocket m fd (TestAddress addr)
+  -> (fd -> m ()) -- ^ Resource cleanup function (socket close)
+  -> (m (Accept m fd (TestAddress addr)) -> m (Either SomeException fd))
+  -- ^ Accepting function
+  -> (Either SomeException fd -> m (Either SomeException fd))
+  -- ^ Assert UniverseState
+  -> m (Either SomeException ())
+runServer localAddress snocket closeF acceptF assertUS = do
+    labelThisThread "server"
+    bracket (open snocket TestFamily)
+            closeF
+            (\fd -> do
+              bind snocket fd localAddress
+              listen snocket fd
+              mbFd <- acceptF (accept snocket fd)
+              mbFd' <- assertUS mbFd
+              traverse_ closeF mbFd'
+              return (void mbFd')
+            )
+
+acceptOne :: MonadMask m => m (Accept m fd addr) -> m (Either SomeException fd)
+acceptOne accept = mask_ $ do
+  accept0 <- accept
+  (accepted, _) <- runAccept accept0
+  case accepted of
+    Accepted fd' _ -> do
+      return (Right fd')
+    AcceptFailure err ->
+      return (Left err)
+
+runClient
+  :: ( MonadThread m
+     , MonadCatch m
+     )
+  => addr -- ^ Local Address
+  -> addr -- ^ Remote Address
+  -> Snocket m fd addr
+  -> (fd -> m ()) -- ^ Resource cleanup function (socket close)
+  -> m (Either SomeException ())
+runClient localAddress remoteAddress snocket closeF = do
+    labelThisThread "client"
+    bracket (openToConnect snocket localAddress)
+            closeF
+            $ \fd -> do
+              bind snocket fd localAddress
+              Right <$> connect snocket fd remoteAddress
+              `catch` (\(e :: SomeException) -> return (Left e))
+
+listenAndConnect
+  :: ( MonadThread m
+     , MonadCatch m
+     , Ord addr
+     )
+  => TestAddress addr -- ^ Local Address
+  -> TestAddress addr -- ^ Remote Address
+  -> Snocket m fd (TestAddress addr)
+  -> m (ObservableNetworkState (TestAddress addr))
+  -> m (Either SomeException ())
+listenAndConnect localAddress remoteAddress snocket getUnivState = do
+    labelThisThread "connectingServer"
+    bracket (open snocket TestFamily)
+            (close snocket)
+            $ \fd -> do
+              bind snocket fd localAddress
+              listen snocket fd
+              bracket (openToConnect snocket localAddress)
+                      (close snocket)
+                      $ \fd' -> do
+                        bind snocket fd' localAddress
+                        res <- (Right <$> connect snocket fd' remoteAddress)
+                              `catch` (\(e :: SomeException) -> return (Left e))
+                        assertUniverseState localAddress remoteAddress
```

Comment:
`assertUniverseState` -> `assertNetworkState`

please `grep -ri universe ouroboros-network-framework` if there are other traces of it still around.